grove/tomography/utils.py

```diff
@@ -0,0 +1,627 @@
+    :param numpy.ndarray confusion_rate_matrix: The matrix of detection probabilities conditional
+    on a prepared qubit state.
+    """
+
+    confusion_rate_matrix = np.asarray(confusion_rate_matrix)
+    if not np.allclose(confusion_rate_matrix.sum(axis=0), np.ones(confusion_rate_matrix.shape[1])):
+        raise CRMUnnormalizedError("Unnormalized confusion matrix:\n{}".format(
+            confusion_rate_matrix))
+    if not (confusion_rate_matrix >= 0).all() or not (confusion_rate_matrix <= 1).all():
+        raise CRMValueError("Confusion matrix must have values in [0, 1]:"
+                            "\n{}".format(confusion_rate_matrix))
+
+    ops = [sum((pi_j * pjk for (pi_j, pjk) in izip(pi_basis.ops, pjs)), 0)
+           for pjs in confusion_rate_matrix]
+    return DiagonalPOVM(pi_basis=pi_basis, confusion_rate_matrix=confusion_rate_matrix, ops=ops)
+
+
+class OperatorBasis(object):
+    """
+    Encapsulate a complete set of basis operators.
+    """
+
+    def __init__(self, labels_ops):
+        """
+        Encapsulates a set of linearly independent operators.
+
+        :param (list|tuple) labels_ops: Sequence of tuples (label, operator) where label is a string
+            and operator a qutip.Qobj operator representation.
+        """
+        self.ops_by_label = OrderedDict(labels_ops)
+        self.labels = list(self.ops_by_label.keys())
+        self.ops = list(self.ops_by_label.values())
+        self.dim = len(self.ops)
+
+        # the basis change transformation matrix from a representation in the operator basis
+        # to the original basis. We enforce CSR sparse matrix representation to have efficient
+        # matrix vector products.
+        self.basis_transform = sphstack([qt.operator_to_vector(opj).data
+                                         for opj in self.ops]).tocsr()
+        self._metric = None
+        self._is_orthonormal = None
+        self._all_hermitian = None
+
+    def metric(self):
+        """
+        Compute a matrix of Hilbert-Schmidt inner products for the basis operators.
+        """
+        if self._metric is None:
+            _log.debug("Computing and caching operator basis metric")
+            self._metric = np.matrix([[(j.dag() * k).tr() for k in self.ops] for j in self.ops])
+        return self._metric
+
+    def is_orthonormal(self):
+        """
+        Compute a matrix of Hilbert-Schmidt inner products for the basis operators.
+        """
+        if self._is_orthonormal is None:
+            _log.debug("Testing and caching if operator basis is orthonormal")
+            self._is_orthonormal = np.allclose(self.metric(), np.eye(self.dim))
+        return self._is_orthonormal
+
+    def all_hermitian(self):
+        """
+        Check if all basis operators are hermitian.
+        """
+        if self._all_hermitian is None:
+            _log.debug("Testing and caching if all basis operator are hermitian")
```

Comment:
In general we don't have log messages in Grove, I don't think... I think it's good to have these here, but let's make sure it's set up right and being put in a log folder.